File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    answer = []
    temp = {}
    
    stages.sort()

    for stage in range(1, N + 1):
        user_count = 0
        unsolved_count = 0
        
        for actual in stages :
            if stage <= actual :
                user_count += 1
            if stage == actual :
                unsolved_count += 1
                
        if user_count == 0:
            temp[stage] = 0  # 분모가 0이면 실패율을 0으로 처리
        else:
            temp[stage] = unsolved_count / user_count
    
    sorted_items_descending = sorted(temp.items(), key=lambda x: x[1], reverse=True)
    
    for key in sorted_items_descending:
        answer.append(key[0])
    
    return answer

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.HashMap;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = {};
        int[] challenger = new int[N+2];
        for(int stage: stages) {
            challenger[stage] += 1;
        }
        
        HashMap<Integer, Double> fails = new HashMap<>();
        double total = stages.length;
        
        for(int i = 1; i<= N; i++) {
            if(challenger[i] == 0) {
                fails.put(i, 0.);
            } else {
                fails.put(i, (double) challenger[i] / total);
                total -= challenger[i];
            }
        }
        
        return fails.entrySet().stream().sorted((o1, o2) -> Double.compare(o2.getValue(), o1.getValue())).mapToInt(HashMap.Entry::getKey).toArray();
    }
}